mantidproject/mantid
# Add absolute normalisation to POLARIS powder diffraction scripts

**Description of work.**

Add a switch to allow an absolute normalisation to be performed during the POLARIS reduction through the ISIS powder diffraction scripts. This will produce a differential cross section with the correct scale. The change was primarily motivated by the total scattering workflow where a Placzek correction is subtracted and therefore the intensity needs to be of the correct scale.

As part of this the interface of some classes used in the workflow have been updated:

- a switch has been added to the CalculatePlaczekSelfScatteringCorrection v1 algorithm to turn off the part of the calculation where the correction size is scaled by a packing fraction. It's not very clear why this scaling was added but it doesn't appear to be required\appropriate if the intensity has been properly normalized

- the setting "crystal_density" has been removed from the SampleDetails class. The density of the material can now be set instead by supplying two of "number_density", "number_density_effective" and "packing_fraction". These three properties on the SampleDetails class are passed through to the Kernel::Material class where they already existed as a result of some work done by ORNL

Finally, there were a couple of glitches relating to the Kernel::Material class that I've fixed up:

- the MayersSampleCorrection algorithm (which is used in the ISIS Powder diffraction scripts) now uses the powder density rather than the full crystal density
- the logic to set the number_density_effective property in SetSample had a bug in it
- you can now supply a chemical formula for an element (eg Si) and a packing_fraction and Mantid calculates the number density and effective number density from these inputs. Previously an incorrect validation prevented this 

**To test:**

Unzip this zip file somewhere: 
[POLARISPowderDiffractionFolderSmall.zip](https://github.com/mantidproject/mantid/files/9103459/POLARISPowderDiffractionFolderSmall.zip)

Update the paths in the file polaris_config_example.yaml so they match where you unzipped the file and also update config_file_path in the script below

Run the following script in workbench. The script creates various workspaces including a workspace group called s_of_q_corrected. There are 5 single spectrum workspaces in here corresponding to the 5 banks on the POLARIS instrument. They should all have y data that is >=1 at all values of Q as a result of this PR. This workspace gets 1 subtracted from it before being passed as input to the PDFFourierTransform algorithm. So this check means the input to PDFFourierTransform is positive:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from isis_powder import polaris, SampleDetails

config_file_path = r"C:\Total Scattering\polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)

polaris.focus(run_number="98533", input_mode='Summed')

q_lim = [[2.5, 3, 4, 6, 7], [3.5, 5, 7, 11, 40]]

polaris.create_total_scattering_pdf(run_number="98533", merge_banks=True,q_lims=q_lim, 
                                    freq_params=[1.4], pdf_type="g(r)", debug=True)
```

Fixes #33606.





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
